File: TravailNo1/Encryption/C/main.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

// Fonction pour vérifier si un caractère est une lettre alphabétique
int is_letter(char c) {
    return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z');
}

// Fonction pour normaliser un caractère en majuscule
char normalize_char(char c) {
    if (c >= 'a' && c <= 'z') {
        return c - 'a' + 'A';
    }
    return c;
}

// Fonction qui réalise le chiffrement/déchiffrement de Vigenère
void vigenere_cipher(FILE *input_file, FILE *output_file, char *key, int encrypt) {
    int key_len = strlen(key);
    int j = 0;  // Pour parcourir la clé

    // Normaliser la clé en majuscules
    for (int i = 0; i < key_len; i++) {
        key[i] = normalize_char(key[i]);
    }

    char c;
    while ((c = fgetc(input_file)) != EOF) {
        if (is_letter(c)) {
            char normalized_c = normalize_char(c);
            int offset = normalized_c - 'A';
            int key_offset = key[j % key_len] - 'A';

            char result;
            if (encrypt) {
                // Chiffrement : additionner l'offset de la clé
                result = (offset + key_offset) % 26 + 'A';
            } else {
                // Déchiffrement : soustraire l'offset de la clé
                result = (offset - key_offset + 26) % 26 + 'A';
            }

            fputc(result, output_file);
            j++;  // Avancer dans la clé seulement pour les lettres
        } else {
            fputc(c, output_file);  // Garder les autres caractères inchangés
        }
    }
}

int main() {
    char key[100];
    int choice;
    char input_file_name[100];
    char output_file_name[100];

    // Demander le nom des fichiers source et de sortie
    printf("Entrez le nom du fichier source : ");
    scanf("%s", input_file_name);

    printf("Entrez le nom du fichier de sortie : ");
    scanf("%s", output_file_name);

    // Demander la clé
    printf("Entrez la clé : ");
    scanf("%s", key);

    // Choisir entre chiffrement et déchiffrement
    printf("Entrez 1 pour chiffrer ou 0 pour déchiffrer : ");
    scanf("%d", &choice);

    // Ouvrir les fichiers
    FILE *input_file = fopen(input_file_name, "r");
    FILE *output_file = fopen(output_file_name, "w");

    if (input_file == NULL || output_file == NULL) {
        printf("Erreur lors de l'ouverture des fichiers.\n");
        return 1;
    }

    // Appliquer le chiffrement ou déchiffrement
    vigenere_cipher(input_file, output_file, key, choice);

    // Fermer les fichiers
    fclose(input_file);
    fclose(output_file);

    printf("Opération terminée avec succès.\n");

    return 0;
}

File: TravailNo1/Encryption/python/main.py
def normalize_char(c):
    # Convertir les caractères en majuscules et ignorer les accents
    if 'a' <= c <= 'z':
        return chr(ord(c) - ord('a') + ord('A'))
    return c

def is_letter(c):
    # Vérifie si un caractère est une lettre alphabétique
    return 'A' <= c <= 'Z' or 'a' <= c <= 'z'

def vigenere_cipher(text, key, encrypt=True):
    result = []
    key_len = len(key)
    j = 0  # Index pour parcourir la clé

    for c in text:
        if is_letter(c):
            normalized_c = normalize_char(c)
            normalized_key = normalize_char(key[j % key_len])

            offset = ord(normalized_c) - ord('A')
            key_offset = ord(normalized_key) - ord('A')

            if encrypt:
                new_char = chr((offset + key_offset) % 26 + ord('A'))
            else:
                new_char = chr((offset - key_offset + 26) % 26 + ord('A'))

            result.append(new_char)
            j += 1
        else:
            result.append(c)

    return ''.join(result)

def process_file(input_filename, output_filename, key, encrypt=True):
    try:
        with open(input_filename, 'r') as infile, open(output_filename, 'w') as outfile:
            text = infile.read()
            result = vigenere_cipher(text, key, encrypt)
            outfile.write(result)
            print(f"Opération terminée avec succès. Résultat écrit dans {output_filename}.")
    except FileNotFoundError:
        print(f"Le fichier {input_filename} n'a pas été trouvé.")
    except IOError as e:
        print(f"Erreur lors de l'accès au fichier : {e}")

def main():
    input_filename = input("Entrez le nom du fichier source : ")
    output_filename = input("Entrez le nom du fichier de sortie : ")
    key = input("Entrez la clé : ")

    choice = input("Entrez 1 pour chiffrer ou 0 pour déchiffrer : ")
    encrypt = choice == '1'

    process_file(input_filename, output_filename, key, encrypt)

if __name__ == "__main__":
    main()
